langlib/lang.query/src/main/ballerina/types.bal

```diff
@@ -650,24 +650,26 @@ class _GroupByFunction {
     }
 
     public function process() returns _Frame|error? {
-        if (self.groupedStream is ()) {
-            _StreamFunction pf = <_StreamFunction>self.prevFunc;
-            _Frame? f = check pf.process();
-            while f is _Frame {
-                anydata & readonly key = (check self.getKey(f)).cloneReadOnly();
-                if self.tbl.hasKey(key) {
-                    self.tbl.get(key).frames.push(f);
-                } else {
-                    self.tbl.add({groupingKey: key, frames: [f]});
+        lock {
```

Comment:
Do we need to add lock for the entire logic?

Reply:
This wraps many shared class level fields.. `groupedStream` and `tbl`.